File: tianshu-ts/src/core/rag/service.ts
import * as fs from "node:fs";
import * as path from "node:path";
import type { BaseProvider } from "../llm/types.js";
import type { Hit } from "./retrieve.js";
import { RagRetriever } from "./retrieve.js";
import { allChunks, getDoc, ingestText, listDocs } from "./store.js";

let _provider: BaseProvider | null = null;

export function setProvider(p: BaseProvider | null): void {
  _provider = p;
}

export function ragProvider(): BaseProvider | null {
  return _provider;
}

async function chatText(p: BaseProvider, prompt: string): Promise<string> {
  const result = await p.chat([{ role: "user", content: prompt }]);
  return (result.content ?? "").trim();
}

async function llmOptional(prompt: string, fallback: string): Promise<string> {
  if (!_provider) return fallback;
  try {
    return (await chatText(_provider, prompt)) || fallback;
  } catch {
    return fallback;
  }
}

export async function rewriteQueries(query: string): Promise<string[]> {
  const rewritten = await llmOptional(
    "你是检索查询改写器。将下面用户查询改写为 1~3 个更利于全文检索的查询,每行一个,只输出查询内容,不要编号。\n查询:" + query,
    query,
  );
  const lines = rewritten.split("\n").map((ln) => ln.trim()).filter(Boolean);
  const unique = [query];
  for (const ln of lines) {
    if (!unique.includes(ln)) unique.push(ln);
  }
  return unique.slice(0, 3);
}

export async function hydeDocs(query: string): Promise<string[]> {
  const fake = await llmOptional(
    "为下面查询写一段 50 词的假设性标准答案短文,直接输出内容。\n查询:" + query,
    "",
  );
  if (!fake || !_provider) return [];
  return [fake];
}

async function titles(docIds: string[]): Promise<string> {
  const parts: string[] = [];
  for (const did of [...new Set(docIds)]) {
    const doc = getDoc(did);
    if (doc && doc.versions.length) {
      const latest = doc.versions[doc.versions.length - 1];
      parts.push(`- ${doc.title} (v${latest.version}, ${latest.ts})`);
    }
  }
  return parts.join("\n");
}

export async function buildAnswer(query: string, hits: Hit[], withHyde = false): Promise<string> {
  const contexts = hits.map(
    (h, i) => `[引用 ${i + 1}][doc:${h.chunk.doc_id} v${h.chunk.version}] ${h.chunk.text}`,
  );
  if (withHyde && _provider) {
    const hyde = await hydeDocs(query);
    contexts.unshift(...hyde);
  }
  const refs = await titles(hits.map((h) => h.chunk.doc_id));
  const prompt =
    "你是知识库问答引擎。仅依据下方引用资料回答用户问题;若资料不足,明确说明缺失信息,不得编造。" +
    "回答中用 [引用 N] 标注依据的引用编号。\n\n引用资料:\n" +
    contexts.join("\n\n") +
    `\n\n问题:${query}\n\n答案:`;
  const answer = await llmOptional(prompt, "参考以上资料生成回答");
  return refs ? `${answer}\n\n参考版本:\n${refs}` : answer;
}

export interface RagQueryResult {
  query: string;
  rewrites: string[];
  hits: Array<{ doc: string; version: number; excerpt: string; score: number }>;
  answer: string;
}

export async function ragQuery(query: string, topK = 5, useHyde = false): Promise<RagQueryResult> {
  const queries = await rewriteQueries(query);
  const retriever = new RagRetriever(allChunks());
  const merged = new Map<string, Hit>();
  for (const q of queries) {
    for (const hit of retriever.search(q, topK)) {
      const key = hit.chunk.doc_id;
      const existing = merged.get(key);
      if (!existing || hit.score > existing.score) merged.set(key, hit);
    }
  }
  const hits = [...merged.values()].sort((a, b) => b.score - a.score).slice(0, topK);
  let answer = "";
  if (hits.length) {
    answer = await buildAnswer(query, hits, useHyde);
  } else {
    answer = "文档库中没有相关内容,请先通过 document_ingest 导入资料。";
  }
  return {
    query,
    rewrites: queries,
    hits: hits.map((h) => ({
      doc: h.chunk.doc_id,
      version: h.chunk.version,
      excerpt: h.chunk.text.slice(0, 200),
      score: Math.round(h.score * 10000) / 10000,
    })),
    answer,
  };
}

export function ingestFile(filePath: string, docId = "", title = ""): string {
  const p = path.resolve(filePath);
  if (!fs.existsSync(p) || !fs.statSync(p).isFile()) {
    throw new Error(`文件不存在: ${filePath}`);
  }
  const text = fs.readFileSync(p, "utf-8");
  const result = ingestText(text, p, docId, title || path.basename(p));
  return `已入库 ${path.basename(p)} → ${result}(${text.length} 字符)`;
}

export function ragDocsList(): string {
  return listDocs();
}

File: tianshu/core/rag/service.py
from __future__ import annotations

from typing import Any

from tianshu.core.llm.base import BaseProvider, LLMMessage
from tianshu.core.rag.retrieve import Hit, RagRetriever, all_chunks
from tianshu.core.rag.store import get_doc, ingest_text, list_docs

_PROVIDER: BaseProvider | None = None


def set_provider(p: BaseProvider | None) -> None:
    global _PROVIDER
    _PROVIDER = p


def provider() -> BaseProvider | None:
    return _PROVIDER


async def _chat_text(p: BaseProvider, prompt: str) -> str:
    result = await p.chat([LLMMessage(role="user", content=prompt)])
    return (result.content or "").strip()


async def _llm_optional(prompt: str, fallback: str) -> str:
    if _PROVIDER is None:
        return fallback
    try:
        return await _chat_text(_PROVIDER, prompt) or fallback
    except Exception:  # noqa: BLE001
        return fallback


async def rewrite_queries(query: str) -> list[str]:
    rewritten = await _llm_optional(
        "你是检索查询改写器。将下面用户查询改写为 1~3 个更利于全文检索的查询,每行一个,只输出查询内容,不要编号。\n查询:" + query,
        query,
    )
    lines = [ln.strip() for ln in rewritten.splitlines() if ln.strip()]
    unique = [query]
    for ln in lines:
        if ln not in unique:
            unique.append(ln)
    return unique[:3]


async def hyde_docs(query: str) -> list[str]:
    fake = await _llm_optional(
        "为下面查询写一段 50 词的假设性标准答案短文,直接输出内容。\n查询:" + query,
        "",
    )
    if not fake:
        return []
    return [fake] if _PROVIDER is not None else []


async def _titles(doc_ids: list[str]) -> str:
    parts = []
    for did in dict.fromkeys(doc_ids):
        doc = get_doc(did)
        if doc and doc.latest:
            parts.append(f"- {doc.title} (v{doc.latest.version}, {doc.latest.ts})")
    return "\n".join(parts)


async def build_answer(query: str, hits: list[Hit], with_hyde: bool = False) -> str:
    contexts = [f"[引用 {i + 1}][doc:{h.chunk.doc_id} v{h.chunk.version}] {h.chunk.text}" for i, h in enumerate(hits)]
    if with_hyde and _PROVIDER is not None:
        hyde = await hyde_docs(query)
        contexts = hyde + contexts
    doc_ids = [h.chunk.doc_id for h in hits]
    refs = await _titles(doc_ids)
    prompt = (
        "你是知识库问答引擎。仅依据下方引用资料回答用户问题;若资料不足,明确说明缺失信息,不得编造。"
        "回答中用 [引用 N] 标注依据的引用编号。\n\n引用资料:\n"
        + "\n\n".join(contexts)
        + f"\n\n问题:{query}\n\n答案:"
    )
    answer = await _llm_optional(prompt, "参考以上资料生成回答")
    return f"{answer}\n\n参考版本:\n{refs}" if refs else answer


async def rag_query(query: str, top_k: int = 5, use_hyde: bool = False) -> dict[str, Any]:
    queries = await rewrite_queries(query)
    retriever = RagRetriever(all_chunks())
    merged: dict[str, Hit] = {}
    for q in queries:
        for hit in retriever.search(q, top_k=top_k):
            key = hit.chunk.doc_id
            if key not in merged or hit.score > merged[key].score:
                merged[key] = hit
    hits = sorted(merged.values(), key=lambda h: h.score, reverse=True)[:top_k]
    answer = "" if hits else "文档库中没有相关内容,请先通过 document_ingest 导入资料。"
    if hits:
        answer = await build_answer(query, hits, with_hyde=use_hyde)
    return {
        "query": query,
        "rewrites": queries,
        "hits": [{"doc": h.chunk.doc_id, "version": h.chunk.version, "excerpt": h.chunk.text[:200], "score": round(h.score, 4)} for h in hits],
        "answer": answer,
    }


async def ingest_file(path: str, doc_id: str = "", title: str = "") -> str:
    from pathlib import Path

    p = Path(path).expanduser().resolve()
    if not p.is_file():
        raise FileNotFoundError(f"文件不存在: {path}")
    text = p.read_text(encoding="utf-8", errors="ignore")
    result = ingest_text(text, str(p), doc_id=doc_id, title=title or p.name)
    return f"已入库 {p.name} → {result}({len(text)} 字符)"



def rag_docs_list() -> str:
    return list_docs()